src/section_list_cache.rs

```diff
@@ -0,0 +1,159 @@
+// Licenses can be found in the root directory of this project at LICENSE, COPYING and CONTRIBUTOR.
+//
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.
+//
+// Please review the Licences for the specific language governing permissions and limitations
+// relating to use of the SAFE Network Software.
+
+use id::PublicId;
+use itertools::Itertools;
+use messages::GroupList;
+use routing_table::Prefix;
+use rust_sodium::crypto::sign::Signature;
+use std::collections::HashMap;
+use super::QUORUM;
+use super::XorName;
+
+pub type Signatures = HashMap<PublicId, Signature>;
+pub type PrefixMap<T> = HashMap<Prefix<XorName>, T>;
+
+pub struct SectionListCache {
+    // all signatures for a group list for a given prefix
+    signatures: PrefixMap<HashMap<GroupList, Signatures>>,
+    // group lists signed by a given public id
+    signed_by: HashMap<PublicId, PrefixMap<GroupList>>,
+    // the latest group list for each prefix with a quorum of signatures
+    lists_cache: PrefixMap<(GroupList, Signatures)>,
+}
+
+impl SectionListCache {
+    pub fn new() -> SectionListCache {
+        SectionListCache {
+            signatures: Default::default(),
+            signed_by: Default::default(),
+            lists_cache: Default::default(),
+        }
+    }
+
+    fn prune(&mut self) {
+        let mut to_remove = vec![];
+        for (prefix, map) in &mut self.signatures {
+            // prune group lists with 0 signatures
+            let lists_to_remove = map.iter()
+                .filter(|&(_, sigs)| sigs.is_empty())
+                .map(|(list, _)| list.clone())
+                .collect_vec();
+            for list in lists_to_remove {
+                let _ = map.remove(&list);
+            }
+            if map.is_empty() {
+                to_remove.push(*prefix);
+            }
+        }
+        // prune prefixes with no group lists
+        for prefix in to_remove {
+            let _ = self.signatures.remove(&prefix);
+        }
+
+        let to_remove = self.signed_by
+            .iter()
+            .filter(|&(_, map)| map.is_empty())
+            .map(|(pub_id, _)| *pub_id)
+            .collect_vec();
+        // prune pub_ids signing nothing
+        for pub_id in to_remove {
+            let _ = self.signed_by.remove(&pub_id);
+        }
+    }
+
+    fn update_lists_cache(&mut self, our_section_size: usize) {
+        for (prefix, map) in &self.signatures {
+            // find the entries with the most signatures
+            let entries = map.iter()
+                .map(|(list, sigs)| (list, sigs.len()))
+                .sorted_by(|lhs, rhs| rhs.1.cmp(&lhs.1));
+            if let Some(&entry) = entries.first() {
+                // entry.0 = list, entry.1 = num of signatures
```

Comment:
You could destructure it, then that comment wouldn't be necessary: `let Some(&(ref list, ref sig_count))`